src/algebra/ring/basic.lean

```diff
@@ -91,6 +91,21 @@ protected def function.surjective.distrib {S} [distrib R] [has_add S] [has_mul S
 ### Semirings
 -/
 
+/-- A not-necessarily-unital, not-necessarily-associative semiring. -/
+@[protect_proj, ancestor add_comm_monoid distrib mul_zero_class]
+class non_unital_non_assoc_semiring (α : Type u) extends
+  add_comm_monoid α, distrib α, mul_zero_class α
+
+/-- An associative but not-necessarily unital semiring. -/
+@[protect_proj, ancestor non_unital_non_assoc_semiring semigroup_with_zero]
+class non_unital_semiring (α : Type u) extends
+  non_unital_non_assoc_semiring α, semigroup_with_zero α
+
+/-- A unital but not-necessarily-associative semiring. -/
+@[protect_proj, ancestor non_unital_non_assoc_semiring mul_one_class]
+class non_assoc_semiring (α : Type u) extends
+  non_unital_non_assoc_semiring α, mul_one_class α
```

Comment:
Something else I've realized is missing - we should have a `non_assoc_semiring` instance on a `subsemiring` of a `non_assoc_semiring`.

Reply:
Given that those entire files will admit generalizations now, perhaps these instances should wait for another PR.

Reply:
We'd also need to make `subsemigroup`, and maybe even `subhas_mul`...?

Reply:
Adding new subobjects is definitely out of scope here, but adding instances of a typeclass on existing subobjects feels in scope (if we want it to be).